File: frontend/src/data/graph.ts
// Superficie del GRAFO de relaciones contra la API real. Vértices (proyección de los `mod_*`) +
// aristas (`relation_edges`). Cada arista lleva su PRODUCTOR (quién la formó) y DOS EJES:
// `provenance` (extracted = leído de la fuente / inferred = el LLM lo dedujo) × `verdict`
// (confirmed/rejected/ambiguous). `label` es la etiqueta canónica derivada de ambos
// (EXTRACTED/INFERRED/AMBIGUOUS/...) y `relation` la justificación corta. Como el resto de
// `@/data`: funciones async + transform snake_case → camelCase.

import { apiGet, apiPost } from "@/lib/api"

export interface GraphNode {
  slug: string
  id: number
  label: string
  kind: string
  sourceInboxIds: number[]
}

export interface GraphEdge {
  id: number
  srcSlug: string
  srcId: number
  dstSlug: string
  dstId: number
  relationType: string
  producer: string
  /** Cómo lo sabemos: "extracted" (determinista) | "inferred" (el LLM lo dedujo). */
  provenance: string
  /** La decisión: "confirmed" | "rejected" | "ambiguous". */
  verdict: string
  /** Etiqueta canónica derivada (EXTRACTED/INFERRED/INFERRED REJECTED/AMBIGUOUS[(inferred)]). */
  label: string
  /** Justificación corta de la relación (la `relation` nombrada del LLM, o texto determinista). */
  relation: string
  /** Groundwork incremental (ADR-021): la arista cambió desde el último mantenimiento. */
  dirty: boolean
  confidence: number | null
  evidence: string
  /** TODOS los mensajes que generaron la pista de co-ocurrencia (no solo el primero del
   * `evidence`); vacío para aristas de otros productores. Mismo drill-down que los nodos. */
  sourceInboxIds: number[]
}

/** Etiqueta canónica de una arista derivada de los dos ejes — espejo de `edges.canonical_label`
 *  (backend). Las aristas reales ya la traen en `label`; las sintéticas del plegado la derivan. */
export function canonicalLabel(provenance: string, verdict: string): string {
  if (verdict === "confirmed") return provenance === "extracted" ? "EXTRACTED" : "INFERRED"
  if (verdict === "rejected") return provenance === "inferred" ? "INFERRED REJECTED" : "REJECTED"
  return provenance === "inferred" ? "AMBIGUOUS (inferred)" : "AMBIGUOUS"
}

export interface GraphData {
  nodes: GraphNode[]
  edges: GraphEdge[]
  /** Medio (email|chat|social) por id de inbox referenciado — para etiquetar «correo/chat/social
   * #N» sin otra llamada; un id ausente cae a «mensaje #N». */
  inboxKinds: Record<number, string>
}

export interface GraphBuildResult {
  cooccurrencePistas: number
  afiliacionReales: number
  pertenenciaReales: number
  contraparteReales: number
  participaReales: number
  chatSenders: number
  canales: number
  highFanoutSkipped: number
}

export interface GraphClusterResult {
  detected: number
  newCandidates: number
  matchedSame: number
  matchedDrift: number
  memoSkipped: number
  deleted: number
  dissolved: number
}

export interface GraphClusterValidateResult {
  blobs: number
  groups: number
  created: number
  synced: number
  dissolved: number
  rejected: number
  promoted: number
  skipped: number
  errors: number
  llmCalls: number
  costUsd: number
}

interface EdgeApiRow {
  id: number
  src_slug: string
  src_id: number
  dst_slug: string
  dst_id: number
  relation_type: string
  producer: string
  provenance: string
  verdict: string
  label: string
  relation: string
  dirty: boolean
  confidence: number | null
  evidence: string
  source_inbox_ids?: number[]
}

interface NodeApiRow {
  slug: string
  id: number
  label: string
  kind: string
  source_inbox_ids: number[]
}

interface GraphApi {
  nodes: NodeApiRow[]
  edges: EdgeApiRow[]
  inbox_kinds?: Record<number, string>
}

function toNode(n: NodeApiRow): GraphNode {
  return { slug: n.slug, id: n.id, label: n.label, kind: n.kind, sourceInboxIds: n.source_inbox_ids ?? [] }
}

function toEdge(e: EdgeApiRow): GraphEdge {
  return {
    id: e.id,
    srcSlug: e.src_slug,
    srcId: e.src_id,
    dstSlug: e.dst_slug,
    dstId: e.dst_id,
    relationType: e.relation_type,
    producer: e.producer,
    provenance: e.provenance,
    verdict: e.verdict,
    label: e.label,
    relation: e.relation,
    dirty: e.dirty,
    confidence: e.confidence,
    evidence: e.evidence,
    sourceInboxIds: e.source_inbox_ids ?? [],
  }
}

/** El grafo del usuario (GET /graph). `verdict` filtra aristas (confirmed|rejected|ambiguous);
 *  `sourceInboxId` ENFOCA el subgrafo en lo que produjo ese mensaje (sus vértices + vecinos). */
export async function fetchGraph(verdict?: string, sourceInboxId?: number): Promise<GraphData> {
  const qs = new URLSearchParams()
  if (verdict) qs.set("verdict", verdict)
  if (sourceInboxId != null) qs.set("source_inbox_id", String(sourceInboxId))
  const suffix = qs.toString() ? `?${qs.toString()}` : ""
  const g = await apiGet<GraphApi>(`/graph${suffix}`)
  return { nodes: g.nodes.map(toNode), edges: g.edges.map(toEdge), inboxKinds: g.inbox_kinds ?? {} }
}

/** Corre el paso determinista (POST /graph/build): materializa pistas + reales sobre lo guardado. */
export async function buildGraph(): Promise<GraphBuildResult> {
  const r = await apiPost<{
    cooccurrence_pistas: number
    afiliacion_reales: number
    pertenencia_reales?: number
    contraparte_reales?: number
    participa_reales?: number
    chat_senders?: number
    canales?: number
    high_fanout_skipped: number
  }>("/graph/build")
  return {
    cooccurrencePistas: r.cooccurrence_pistas,
    afiliacionReales: r.afiliacion_reales,
    pertenenciaReales: r.pertenencia_reales ?? 0,
    contraparteReales: r.contraparte_reales ?? 0,
    participaReales: r.participa_reales ?? 0,
    chatSenders: r.chat_senders ?? 0,
    canales: r.canales ?? 0,
    highFanoutSkipped: r.high_fanout_skipped,
  }
}

/** Detecta los cúmulos (Louvain) y los reconcilia contra lo persistido (POST /graph/cluster, sin
 *  LLM). Idempotente: re-detectar la misma partición no cambia nada. */
export async function clusterGraph(): Promise<GraphClusterResult> {
  const r = await apiPost<{
    detected: number
    new_candidates: number
    matched_same: number
    matched_drift: number
    memo_skipped: number
    deleted: number
    dissolved: number
  }>("/graph/cluster")
  return {
    detected: r.detected,
    newCandidates: r.new_candidates,
    matchedSame: r.matched_same,
    matchedDrift: r.matched_drift,
    memoSkipped: r.memo_skipped,
    deleted: r.deleted,
    dissolved: r.dissolved,
  }
}

/** Parte con el LLM los blobs pendientes (POST /graph/cluster/validate): cada blob → N contextos
 *  (hijos confirmed), promueve sus pistas internas y materializa las aristas `miembro_de`. Usa el
 *  LLM → TIENE COSTO. Solo toca los pendientes. */
export async function validateClusters(): Promise<GraphClusterValidateResult> {
  const r = await apiPost<{
    blobs: number
    groups: number
    created: number
    synced: number
    dissolved: number
    rejected: number
    promoted: number
    skipped: number
    errors: number
    llm_calls: number
    cost_usd: number
  }>("/graph/cluster/validate")
  return {
    blobs: r.blobs,
    groups: r.groups,
    created: r.created,
    synced: r.synced,
    dissolved: r.dissolved,
    rejected: r.rejected,
    promoted: r.promoted,
    skipped: r.skipped,
    errors: r.errors,
    llmCalls: r.llm_calls,
    costUsd: r.cost_usd,
  }
}

export interface GraphConfirmResult {
  edges: number
  confirmedRecibo: number
  messages: number
  chatSkipped: number
  llmConfirmed: number
  llmRejected: number
  llmDejar: number
  gated: number
  summaries: number
  errors: number
  llmCalls: number
  costUsd: number
}

/** Confirmación por-mensaje de las co-ocurrencias ambiguas (POST /graph/confirm): metodología B
 *  (recibo a priori + LLM con compuerta alias-aware). Usa el LLM → TIENE COSTO. Monótono. */
export async function confirmCooccurrences(): Promise<GraphConfirmResult> {
  const r = await apiPost<{
    edges: number
    confirmed_recibo: number
    messages: number
    chat_skipped: number
    llm_confirmed: number
    llm_rejected: number
    llm_dejar: number
    gated: number
    summaries: number
    errors: number
    llm_calls: number
    cost_usd: number
  }>("/graph/confirm")
  return {
    edges: r.edges,
    confirmedRecibo: r.confirmed_recibo,
    messages: r.messages,
    chatSkipped: r.chat_skipped,
    llmConfirmed: r.llm_confirmed,
    llmRejected: r.llm_rejected,
    llmDejar: r.llm_dejar,
    gated: r.gated,
    summaries: r.summaries,
    errors: r.errors,
    llmCalls: r.llm_calls,
    costUsd: r.cost_usd,
  }
}

// --- Cronología / story de un cúmulo (GET /graph/clusters/:id/timeline) --------------------------- //

/** Un suceso fechado del cúmulo. `at` es ISO en hora local; `precision` ∈ datetime|date|inferred. */
export interface TimelineEvent {
  slug: string
  id: number
  kind: string
  label: string
  at: string
  precision: string
  sourceInboxIds: number[]
}

/** Un miembro sin fecha de evento (elenco/contexto: identidad, hábito). */
export interface TimelineActor {
  slug: string
  id: number
  kind: string
  label: string
  sourceInboxIds: number[]
}

export interface ClusterTimelineData {
  cluster: {
    id: number
    name: string
    description: string
    confidence: number | null
    memberCount: number
  }
  events: TimelineEvent[]
  actors: TimelineActor[]
  /** Medio (email|chat|social) por id de inbox referenciado por sucesos/elenco. */
  inboxKinds: Record<number, string>
}

interface TimelineEventApi {
  slug: string
  id: number
  kind: string
  label: string
  at: string
  precision: string
  source_inbox_ids: number[]
}

interface ClusterTimelineApi {
  cluster: {
    id: number
    name: string
    description: string
    confidence: number | null
    member_count: number
  }
  events: TimelineEventApi[]
  actors: Omit<TimelineEventApi, "at" | "precision">[]
  inbox_kinds?: Record<number, string>
}

/** Cronología de un cúmulo CONFIRMADO: sus sucesos fechados (ordenados) + el elenco (sin fecha). */
export async function fetchClusterTimeline(id: number): Promise<ClusterTimelineData> {
  const r = await apiGet<ClusterTimelineApi>(`/graph/clusters/${id}/timeline`)
  return {
    cluster: {
      id: r.cluster.id,
      name: r.cluster.name,
      description: r.cluster.description,
      confidence: r.cluster.confidence,
      memberCount: r.cluster.member_count,
    },
    events: r.events.map((e) => ({
      slug: e.slug,
      id: e.id,
      kind: e.kind,
      label: e.label,
      at: e.at,
      precision: e.precision,
      sourceInboxIds: e.source_inbox_ids ?? [],
    })),
    actors: r.actors.map((a) => ({
      slug: a.slug,
      id: a.id,
      kind: a.kind,
      label: a.label,
      sourceInboxIds: a.source_inbox_ids ?? [],
    })),
    inboxKinds: r.inbox_kinds ?? {},
  }
}

File: src/memex/api/routers/graph.py
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import text
from sqlalchemy.engine import Connection

from memex.api.auth import current_user_id
from memex.api.schemas import (
    GraphBuildResult,
    GraphClusterResult,
    GraphClustersResponse,
    GraphClusterTimeline,
    GraphClusterValidateResult,
    GraphResponse,
)
from memex.config import settings
from memex.db import connection
from memex.logging import get_logger
from memex.relations.clusters_llm import run_cluster_partition
from memex.relations.decisions import edge_sources
from memex.relations.deterministic import build_relations, vertex_inbox_ids
from memex.relations.edges import list_edges
from memex.relations.reconcile import detect_and_reconcile
from memex.relations.timeline import cluster_timeline
from memex.relations.vertices import list_vertices
from memex.sources import kind_for_type

router = APIRouter(prefix="/graph", tags=["graph"])

UserID = Annotated[int, Depends(current_user_id)]

_log = get_logger("memex.api.graph")


def _inbox_kinds(conn: Connection, user_id: int, inbox_ids: set[int]) -> dict[int, str]:
    """Medio (email|chat|social) de cada mensaje referenciado: inbox → sources.type → SourceKind.
    Tipos sin SourceKind registrada se omiten (el front cae a «mensaje #N»)."""
    if not inbox_ids:
        return {}
    rows = conn.execute(
        text(
            "SELECT i.id AS id, s.type AS type FROM inbox i "
            "JOIN sources s ON s.id = i.source_id "
            "WHERE i.user_id = :u AND i.id = ANY(:ids)"
        ),
        {"u": user_id, "ids": sorted(inbox_ids)},
    ).mappings()
    kinds: dict[int, str] = {}
    for r in rows:
        try:
            kinds[int(r["id"])] = kind_for_type(str(r["type"])).value
        except KeyError:  # tipo sin kind (p. ej. un push custom) → el front usa el fallback
            continue
    return kinds


@router.get("", response_model=GraphResponse)
async def get_graph(
    user_id: UserID,
    status: Annotated[str | None, Query(description="pista|confirmed|rejected")] = None,
    source_inbox_id: Annotated[
        int | None,
        Query(description="enfoca: solo los vértices producidos por este mensaje + sus vecinos"),
    ] = None,
) -> dict[str, Any]:
    """El grafo del user: vértices (proyectados de las tablas `mod_*`) + aristas (`relation_edges`).

    Un vértice se direcciona por `(slug, id)`; inbox NO es vértice (es procedencia, drill-down):
    cada nodo lleva sus `source_inbox_ids` (los mensajes de los que salió) para abrir el mensaje
    original desde el grafo, y `inbox_kinds` trae el medio (email|chat|social) de cada uno. Las
    aristas llevan su `producer` y su `status` (`pista`/`confirmed`); el front filtra por `status`.
    Con `source_inbox_id` el grafo se ENFOCA en lo que produjo ese mensaje (sentido inverso del
    drill-down nodo→mensaje): sus vértices + los vecinos a un salto. Solo LECTURA: no dispara el
    armado (POST /graph/build).
    """
    with connection() as conn:
        verts = list_vertices(conn, user_id)
        edges = list_edges(conn, user_id, status=status)
        prov = vertex_inbox_ids(conn, user_id)
        if source_inbox_id is not None:
            # Vértices producidos por ESE mensaje (semilla) + las aristas que los tocan; se
            # conservan los vecinos al otro extremo para no dejar aristas colgando. Mensaje sin
            # nada → grafo vacío.
            seed = {ref for ref, ids in prov.items() if source_inbox_id in ids}
            edges = [e for e in edges if e.src in seed or e.dst in seed]
            keep = set(seed)
            for e in edges:
                keep.add(e.src)
                keep.add(e.dst)
            verts = [v for v in verts if v.ref in keep]
        # Poda de aristas huérfanas: ambos extremos deben ser vértices PRESENTES (vivos en default,
        # dentro del subgrafo en foco). Cruza-filtra contra el set final de `verts` → nunca se sirve
        # una arista a un nodo ausente (consolidado tombstoneado / fila borrada / merge sin GC aún).
        present = {v.ref for v in verts}
        edges = [e for e in edges if e.src in present and e.dst in present]
        # Procedencia por ARISTA (relation_edge_sources): todos los mensajes que generaron cada
        # pista de co-ocurrencia — el drill-down de aristas, espejo del de nodos.
        edge_srcs = edge_sources(conn, [e.id for e in edges])
        # El medio de cada mensaje referenciado por el set FINAL de vértices Y aristas (por eso se
        # calcula acá adentro, después del foco/poda).
        referenced: set[int] = set()
        for v in verts:
            referenced |= prov.get(v.ref, set())
        for ids in edge_srcs.values():
            referenced |= ids
        inbox_kinds = _inbox_kinds(conn, user_id, referenced)
    nodes = [
        {
            "slug": v.slug,
            "id": v.id,
            "label": v.label,
            "kind": v.kind,
            "source_inbox_ids": sorted(prov.get(v.ref, set())),
        }
        for v in verts
    ]
    out_edges = [
        {
            "id": e.id,
            "src_slug": e.src.slug,
            "src_id": e.src.id,
            "dst_slug": e.dst.slug,
            "dst_id": e.dst.id,
            "relation_type": e.relation_type,
            "producer": e.producer,
            "status": e.status,
            "confidence": float(e.confidence) if e.confidence is not None else None,
            "evidence": e.evidence,
            "source_inbox_ids": sorted(edge_srcs.get(e.id, set())),
        }
        for e in edges
    ]
    _log.info(
        "graph.read",
        user_id=user_id,
        nodes=len(nodes),
        edges=len(out_edges),
        inbox_kinds=len(inbox_kinds),
    )
    return {"nodes": nodes, "edges": out_edges, "inbox_kinds": inbox_kinds}


@router.post("/build", response_model=GraphBuildResult)
async def build_graph(user_id: UserID) -> dict[str, Any]:
    """Corre el paso de relaciones DETERMINISTAS (on-demand, explícito): materializa pistas de
    co-ocurrencia + afiliaciones reales sobre lo ya guardado. Idempotente y sin LLM. NO dispara
    extracción/consolidación (consume lo disponible). El tope de fan-out es configurable
    (`MEMEX_COOCCURRENCE_CAP`)."""
    with connection() as conn:
        stats = build_relations(conn, user_id, cooccurrence_cap=settings.cooccurrence_cap)
    _log.info(
        "graph.build.api",
        user_id=user_id,
        pistas=stats.cooccurrence_pistas,
        reales=stats.afiliacion_reales,
        pertenencia=stats.pertenencia_reales,
        contraparte=stats.contraparte_reales,
        cumple=stats.cumple_reales,
        participa=stats.participa_reales,
        canales=stats.canales,
        chat_senders=stats.chat_senders,
        skipped=stats.high_fanout_skipped,
        orphans_pruned=stats.orphans_pruned,
        stale_pruned=stats.stale_pruned,
    )
    return {
        "cooccurrence_pistas": stats.cooccurrence_pistas,
        "afiliacion_reales": stats.afiliacion_reales,
        "pertenencia_reales": stats.pertenencia_reales,
        "contraparte_reales": stats.contraparte_reales,
        "cumple_reales": stats.cumple_reales,
        "participa_reales": stats.participa_reales,
        "high_fanout_skipped": stats.high_fanout_skipped,
        "orphans_pruned": stats.orphans_pruned,
        "stale_pruned": stats.stale_pruned,
        "cluster_edges": stats.cluster_edges,
        "chat_senders": stats.chat_senders,
        "canales": stats.canales,
    }


@router.post("/cluster", response_model=GraphClusterResult)
async def cluster_graph(user_id: UserID) -> dict[str, Any]:
    """Detecta los cúmulos (Louvain) y los reconcilia contra lo persistido. On-demand, SIN LLM e
    idempotente: re-detectar la misma partición no cambia nada. NO dispara el armado del grafo
    (POST /graph/build) ni la validación LLM (POST /graph/cluster/validate)."""
    with connection() as conn:
        stats = detect_and_reconcile(conn, user_id)
    _log.info(
        "graph.cluster.api",
        user_id=user_id,
        detected=stats.detected,
        new_candidates=stats.new_candidates,
        dissolved=stats.dissolved,
    )
    return {
        "detected": stats.detected,
        "matched_same": stats.matched_same,
        "matched_drift": stats.matched_drift,
        "new_candidates": stats.new_candidates,
        "memo_skipped": stats.memo_skipped,
        "deleted": stats.deleted,
        "dissolved": stats.dissolved,
    }


@router.post("/cluster/validate", response_model=GraphClusterValidateResult)
async def validate_clusters(
    user_id: UserID,
    limit: Annotated[int | None, Query(description="máximo de blobs a particionar")] = None,
) -> dict[str, Any]:
    """Parte con el LLM los blobs `candidate`: cada blob → N contextos (hijos confirmed),
    preservando la identidad de los hijos al re-particionar, promoviendo las pistas intra-grupo y
    materializando las aristas `miembro_de`. Usa el LLM (cuesta); on-demand. Solo los pendientes."""
    stats = await run_cluster_partition(user_id, limit=limit)
    _log.info(
        "graph.cluster.partition.api",
        user_id=user_id,
        blobs=stats.blobs,
        groups=stats.groups,
        errors=stats.errors,
    )
    return {
        "blobs": stats.blobs,
        "groups": stats.groups,
        "created": stats.created,
        "synced": stats.synced,
        "dissolved": stats.dissolved,
        "rejected": stats.rejected,
        "promoted": stats.promoted,
        "skipped": stats.skipped,
        "errors": stats.errors,
        "llm_calls": stats.cost.calls,
        "cost_usd": float(stats.cost.cost_usd),
    }


@router.get("/clusters", response_model=GraphClustersResponse)
async def list_clusters(
    user_id: UserID,
    status: Annotated[str | None, Query(description="filtra por estado del cúmulo")] = None,
) -> dict[str, Any]:
    """Lista los cúmulos del user (opcionalmente filtrados por `status`)."""
    sql = (
        "SELECT id, status, name, description, confidence, member_count "
        "FROM relation_clusters WHERE user_id = :u"
    )
    params: dict[str, Any] = {"u": user_id}
    if status is not None:
        sql += " AND status = :st"
        params["st"] = status
    sql += " ORDER BY status, id"
    with connection() as conn:
        rows = conn.execute(text(sql), params).mappings().all()
    clusters = [
        {
            "id": int(r["id"]),
            "status": str(r["status"]),
            "name": str(r["name"]),
            "description": str(r["description"]),
            "confidence": float(r["confidence"]) if r["confidence"] is not None else None,
            "member_count": int(r["member_count"]),
        }
        for r in rows
    ]
    _log.info("graph.clusters.api", user_id=user_id, clusters=len(clusters))
    return {"clusters": clusters}


@router.get("/clusters/{cluster_id}/timeline", response_model=GraphClusterTimeline)
async def cluster_timeline_view(cluster_id: int, user_id: UserID) -> dict[str, Any]:
    """Cronología (story) de un cúmulo CONFIRMADO: sus miembros fechados como sucesos ordenados +
    el elenco (miembros sin fecha: identidades, hábitos). Solo lectura. 404 si no existe / no
    confirmed. Fechas en hora local (America/Bogota) con `precision` para mostrar hora solo si es
    real."""
    with connection() as conn:
        tl = cluster_timeline(conn, user_id, cluster_id)
        inbox_kinds: dict[int, str] = {}
        if tl is not None:
            referenced = {i for e in tl.events for i in e.source_inbox_ids}
            referenced |= {i for a in tl.actors for i in a.source_inbox_ids}
            inbox_kinds = _inbox_kinds(conn, user_id, referenced)
    if tl is None:
        raise HTTPException(status_code=404, detail="cúmulo no encontrado o no confirmado")
    _log.info(
        "graph.cluster.timeline.api",
        user_id=user_id,
        cluster_id=cluster_id,
        events=len(tl.events),
        actors=len(tl.actors),
    )
    return {
        "cluster": {
            "id": tl.cluster.id,
            "name": tl.cluster.name,
            "description": tl.cluster.description,
            "confidence": tl.cluster.confidence,
            "member_count": tl.cluster.member_count,
        },
        "events": [
            {
                "slug": e.slug,
                "id": e.id,
                "kind": e.kind,
                "label": e.label,
                "at": e.at,
                "precision": e.precision,
                "source_inbox_ids": e.source_inbox_ids,
            }
            for e in tl.events
        ],
        "actors": [
            {
                "slug": a.slug,
                "id": a.id,
                "kind": a.kind,
                "label": a.label,
                "source_inbox_ids": a.source_inbox_ids,
            }
            for a in tl.actors
        ],
        "inbox_kinds": inbox_kinds,
    }
